fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/planning/BooleanPredicateNormalizer.java

```diff
@@ -0,0 +1,408 @@
+     * size limit, throw a {@link NormalFormTooLargeException}.
+     * @param predicate the query predicate to be normalized
+     * @return the predicate in disjunctive normal form
+     * @throws NormalFormTooLargeException if the DNF would exceed the size limit
+     */
+    @Nonnull
+    public Optional<QueryPredicate> normalize(@Nullable final QueryPredicate predicate) {
+        return normalize(predicate, true);
+    }
+
+    /**
+     * Convert the given predicate to its respective normal form, if necessary. If the size of the normal form would
+     * exceed the size limit, throw a {@link NormalFormTooLargeException}.
+     * @param predicate the query predicate to be normalized
+     * @param failIfTooLarge indicates whether we should throw an exception if the normalization is too big
+     *        (beyond {@link #sizeLimit}).
+     * @return the predicate in disjunctive normal form
+     * @throws NormalFormTooLargeException if the DNF would exceed the size limit
+     */
+    @Nonnull
+    public Optional<QueryPredicate> normalize(@Nullable final QueryPredicate predicate, boolean failIfTooLarge) {
+        if (!needsNormalize(predicate)) {
+            return Optional.empty();
+        } else if (!shouldNormalize(predicate)) {
+            if (failIfTooLarge) {
+                throw new NormalFormTooLargeException(Objects.requireNonNull(predicate));
+            } else {
+                return Optional.empty();
+            }
+        } else {
+            final List<Collection<? extends QueryPredicate>> majorOfMinor = toNormalized(Objects.requireNonNull(predicate), false);
+            trimTerms(majorOfMinor);
+            return Optional.of(mode.majorWithChildren(majorOfMinor.stream().map(mode::minorWithChildren).collect(Collectors.toList())));
+        }
+    }
+
+    private boolean needsNormalize(@Nullable final QueryPredicate predicate) {
+        if (predicate == null) {
+            return false;
+        }
+
+        return isBooleanPredicate(predicate) &&
+               StreamSupport.stream(predicate.getChildren().spliterator(), false).anyMatch(this::isBooleanPredicate);
+    }
+
+    private boolean shouldNormalize(@Nullable final QueryPredicate predicate) {
+        try {
+            return getNormalizedSize(predicate) <= sizeLimit;
+        } catch (ArithmeticException e) {
+            // Our computation caused an integer overflow so the DNF is _definitely_ too big.
+            return false;
+        }
+    }
+
+    private boolean isBooleanPredicate(@Nullable final QueryPredicate predicate) {
+        return predicate instanceof AndOrPredicate ||
+               predicate instanceof NotPredicate;
+    }
+
+    int getNormalizedSize(@Nullable final QueryPredicate predicate) {
+        if (predicate == null) {
+            return 0;
+        }
+        return toNormalizedSize(predicate, false);
+    }
+
+    private int toNormalizedSize(@Nonnull final QueryPredicate predicate, final boolean negate) {
```

Comment:
In case the numbers can grow quite large, this calculation can take a while. Would it be more efficient to take in the limit and short-circuit the count when that limit is reached instead?